documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,227 @@
+.Hot Rod Server worker thread pool size
+----
+<hotrod-connector socket-binding="hotrod" cache-container="local" worker-threads="200">
+   <!-- Additional configuration here -->
+</hotrod-connector>
+----
+
+== Cache Store performance
+
+In order to achieve the best performance, please follow the recommendations below when using Cache Stores:
+
+* Use async mode (write-behind) if possible
+* Prevent cache misses by preloading data
+* For JDBC Cache Store:
+** Use indexes on `id` column to prevent table scans
+** Use PRIMARY_KEY on `id` column
+** Configure batch-size, fetch-size, etc
+
+== Hints for program developers
+
+There are also several hints for developers which can be easily applied to the client application and will boost up the performance.
+
+=== Ignore return values
+
+When you're not interested in returning value of the `#put(k, v)` or `#remove(k)` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
+
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
+----
+Cache noPreviousValueCache = cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES);
+noPreviousValueCache.put(k, v);
+----
+
+=== Use `Externalizable` for marshalling
+
+Infinispan uses JBoss Marshalling to transfer objects over the wire. The most efficient way to marshall user data is to provide an AdvancedExternalizer.
+This solutions prevents JBoss Marshalling from sending class name over the network and allows to save some bandwidth:
+
+[source,java]
+.User entity with Externalizer
+----
+import org.infinispan.marshall.AdvancedExternalizer;
+
+public class Book {
+
+   final String name;
+   final String author;
+
+   public Book(String name, String author) {
+      this.name = name;
+      this.author = author;
+   }
+
+   public static class BookExternalizer
+            implements AdvancedExternalizer<Book> {
+
+      @Override
+      public void writeObject(ObjectOutput output, Book book)
+            throws IOException {
+         output.writeObject(book.name);
+         output.writeObject(book.author);
+      }
+
+      @Override
+      public Person readObject(ObjectInput input)
+            throws IOException, ClassNotFoundException {
+         return new Person((String) input.readObject(), (String) input.readObject());
+      }
+
+      @Override
+      public Set<Class<? extends Book>> getTypeClasses() {
+         return Util.<Class<? extends Book>>asSet(Book.class);
+      }
+
+      @Override
+      public Integer getId() {
+         return 2345;
+      }
+   }
+}
+----
+
+The Externalizer must be registered in cache configuration. See configuration examples below:
+
+[source,java]
+.Adding Externalizer using XML
+----
+<cache-container>
+   <serialization>
+      <advanced-externalizer class="Book$BookExternalizer"/>
+   </serialization>
+</cache-container>
+----
+
+[source,java]
+.Adding Externalizer using Java
+----
+GlobalConfigurationBuilder builder = ...
+builder.serialization().addAdvancedExternalizer(new Book.BookExternalizer());
+----
+
```

Comment:
Maybe you could add a link to the docu on externalizers if the reader is interested in more details?